libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -226,87 +350,93 @@ namespace pika::mpi::experimental {
-                // we don't notify any condition either
-                return;
-            }
-            add_to_request_callback_queue(request_callback{request, PIKA_MOVE(callback), stream});
+            add_to_request_callback_queue(request_callback{request, PIKA_MOVE(callback)});
+            return true;
         }
 
+        // -------------------------------------------------------------
         // an MPI error handling type that we can use to intercept
         // MPI errors if we enable the error handler
         MPI_Errhandler pika_mpi_errhandler = 0;
 
+        // -------------------------------------------------------------
         // function that converts an MPI error into an exception
         void pika_MPI_Handler(MPI_Comm*, int* errorcode, ...)
         {
-            mpi_debug.debug(debug::detail::str<>("pika_MPI_Handler"));
-            PIKA_THROW_EXCEPTION(
-                pika::error::invalid_status, "pika_MPI_Handler", "{}", error_message(*errorcode));
+            PIKA_DETAIL_DP(mpi_debug<5>, debug(str<>("pika_MPI_Handler")));
+            throw mpi_exception(*errorcode, error_message(*errorcode));
         }
 
+        // -------------------------------------------------------------
         // set an error handler for communicators that will be called
         // on any error instead of the default behavior of program termination
         void set_error_handler()
         {
-            mpi_debug.debug(debug::detail::str<>("set_error_handler"));
+            PIKA_DETAIL_DP(mpi_debug<5>, debug(str<>("set_error_handler")));
 
             MPI_Comm_create_errhandler(detail::pika_MPI_Handler, &detail::pika_mpi_errhandler);
             MPI_Comm_set_errhandler(MPI_COMM_WORLD, detail::pika_mpi_errhandler);
         }
 
+        bool poll_request(MPI_Request req)
+        {
+            int flag;
+            MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
+            if (flag)
+            {
+                PIKA_DETAIL_DP(mpi_debug<5>, debug(str<>("poll MPI_Test ok"), ptr(req)));
+            }
+            return flag;
+        }
+
+        // -------------------------------------------------------------
         /// Remove all entries in request and callback vectors that are invalid
         /// Ideally, would use a zip iterator to do both using remove_if
         void compact_vectors()
         {
-            size_t const size = detail::mpi_data_.request_vector_.size();
+            using detail::mpi_data_;
+            /*
+            auto z = ranges::views::zip(mpi_data_.requests_, mpi_data_.callbacks_);
+            auto e = ranges::remove_if(z, [](auto&& r) { return r.first == MPI_REQUEST_NULL; });
+            mpi_data_.requests_.erase(
+                mpi_data_.requests_.begin() + (e - z.begin()), mpi_data_.requests_.end());
+            mpi_data_.callbacks_.erase(
+                mpi_data_.callbacks_.begin() + (e - z.begin()), mpi_data_.callbacks_.end());
+            */
```

Comment:
```suggestion
```

Reply:
grumble grumble. that's useful code that I will want when ranges are available.

Reply:
Just FYI, we'd have to wait for C++23 to get `zip`. We can of course consider using standalone ranges-v3 until then but we should see how many use cases we have for it before adding it as a dependency.